File: C++/lintcode_cpp/521_remove_duplicate_numbers_in_array.cpp
class Solution {
public:
    /**
     * @param nums: an array of integers
     * @return: the number of unique integers
     */
    int deduplication(vector<int> &nums) {
        // write your code here
        if (nums.size() < 2) return nums.size();
        sort(nums.begin(), nums.end());
        int idx = 1;
        for (int i = 1; i < nums.size(); i++) {
            if (nums[i] != nums[i-1]) {
                nums[idx] = nums[i];
                idx++;
            }
        }
        return idx;
    }
};

File: Python/lintcode_python/521_remove_duplicate_numbers_in_array.py
# solution 1, O(n), with extra space
# class Solution:
#     """
#     @param: nums: an array of integers
#     @return: the number of unique integers
#     """
#     def deduplication(self, nums):
#         seen = set()
#         idx = 0
        
#         for i in range(len(nums)):
#             if nums[i] not in seen:
#                 seen.add(nums[i])
#                 nums[idx] = nums[i]
#                 idx += 1;
#         return idx;

# solution 2, O(nlogn) without extra space
class Solution:
    """
    @param: nums: an array of integers
    @return: the number of unique integers
    """
    def deduplication(self, nums):
        if not nums:
            return 0
        nums.sort()
        idx = 1
        
        for i in range(1, len(nums)):
            if nums[i] != nums[i-1]:
                nums[idx] = nums[i]
                idx += 1
        return idx
